File: PycharmProjects/ta/strategy.py
__author__ = 'kocat_000'

import shared.directory_names as dn
import shared.calendar_utilities as cu
import os.path
import pandas as pd
import numpy as np
import shared.converters as conv
import datetime as dt
import my_sql_routines.my_sql_utilities as msu
import contract_utilities.expiration as exp
import get_price.get_futures_price as gfp
import get_price.get_options_price as gop
import time as tm
pd.options.mode.chained_assignment = None

def generate_db_strategy_from_strategy_sheet(**kwargs):

    id = kwargs['id']
    strategy_class = kwargs['strategy_class']

    strategy_output = load_strategy_file(**kwargs)

    del strategy_output['z5']
    del strategy_output['z6']
    del strategy_output['z7']

    strategy = strategy_output.loc[id]

    if strategy_class == 'futures_butterfly':
        if strategy['Q'] <= 49:
            strategy_alias = strategy['tickerHead'] + strategy['ticker1'][-5:] + strategy['ticker2'][-5:] + \
                             strategy['ticker2'][-5:] + strategy['ticker3'][-5:]
        elif strategy['Q'] >= 51:
            strategy_alias = strategy['tickerHead'] + strategy['ticker2'][-5:] + strategy['ticker3'][-5:] + \
                       strategy['ticker1'][-5:] + strategy['ticker2'][-5:]

    description_string = 'strategy_class=' + strategy_class + '&' + \
                          conv.convert_from_dictionary_to_string(dictionary_input=strategy)

    return generate_db_strategy_from_alias(alias=strategy_alias, description_string=description_string)


def get_net_position_4strategy_alias(**kwargs):

    alias = kwargs['alias']

    if 'as_of_date' in kwargs.keys():
        as_of_date = kwargs['as_of_date']
    else:
        as_of_date = exp.doubledate_shift_bus_days()

    con = msu.get_my_sql_connection(**kwargs)
    trades_frame = get_trades_4strategy_alias(alias=alias,con=con)

    as_of_datetime = cu.convert_doubledate_2datetime(as_of_date)

    trades_frame = trades_frame[trades_frame['trade_date'] <= as_of_datetime]

    trades_frame['full_ticker'] = [trades_frame['ticker'].iloc[x] if trades_frame['instrument'].iloc[x] == 'F' else
                                   trades_frame['ticker'].iloc[x] + '_' + trades_frame['option_type'].iloc[x] + str(trades_frame['strike_price'].iloc[x])
                                   for x in range(len(trades_frame.index))]

    grouped = trades_frame.groupby('full_ticker')

    net_position = pd.DataFrame()

    net_position['ticker'] = (grouped['ticker'].first()).values
    net_position['option_type'] = (grouped['option_type'].first()).values
    net_position['strike_price'] = (grouped['strike_price'].first()).values
    net_position['instrument'] = (grouped['instrument'].first()).values
    net_position['qty'] = (grouped['trade_quantity'].sum()).values

    net_position['qty'] = net_position['qty'].round(2)

    if 'con' not in kwargs.keys():
        con.close()

    return net_position[net_position['qty'] != 0]


def generate_db_strategy_from_alias(**kwargs):

    alias = kwargs['alias']
    description_string = kwargs['description_string']

    con = msu.get_my_sql_connection(**kwargs)
    cur = con.cursor()

    for i in range(1, 50):

        if i > 1:
            alias_modified = alias + '_' + str(i)
        else:
            alias_modified = alias

        strategy_id = get_strategy_id_from_alias(alias=alias_modified, con=con)

        if not strategy_id:
            break

    now_date = dt.datetime.now().date()

    column_str = "alias, open_date, close_date, created_date, last_updated_date, description_string"
    insert_str = ("%s, " * 6)[:-2]

    tuple_to_load = (alias_modified, now_date, cu.convert_doubledate_2datetime(30000101),
                     now_date, now_date, description_string)

    final_str = "INSERT INTO strategy (%s) VALUES (%s)" % (column_str, insert_str)
    cur.execute(final_str,tuple_to_load)
    con.commit()

    strategy_id = get_strategy_id_from_alias(alias=alias_modified, con=con)

    if 'con' not in kwargs.keys():
        con.close()

    return {'alias': alias_modified, 'strategy_id': strategy_id}


def get_trades_4strategy_alias(**kwargs):

    con = msu.get_my_sql_connection(**kwargs)

    sql_query = 'SELECT tr.id, tr.ticker, tr.option_type, tr.strike_price, tr.trade_price, tr.trade_quantity, tr.trade_date, tr.instrument, tr.real_tradeQ ' + \
                 'FROM strategy as str INNER JOIN trades as tr ON tr.strategy_id=str.id ' + \
                 'WHERE str.alias=\'' + kwargs['alias'] + '\''
    cur = con.cursor()
    cur.execute(sql_query)
    data = cur.fetchall()

    if 'con' not in kwargs.keys():
        con.close()

    trade_frame = pd.DataFrame(data, columns=['id', 'ticker', 'option_type', 'strike_price', 'trade_price', 'trade_quantity', 'trade_date', 'instrument', 'real_tradeQ'])
    trade_frame['trade_price'] = [float(x) if x is not None else float('NaN') for x in trade_frame['trade_price'].values]
    trade_frame['trade_quantity'] = trade_frame['trade_quantity'].astype('float64')
    trade_frame['strike_price'] = trade_frame['strike_price'].astype('float64')
    return trade_frame


def get_strategy_id_from_alias(**kwargs):
    alias = kwargs['alias']
    con = msu.get_my_sql_connection(**kwargs)
    cur = con.cursor()

    query_string = 'SELECT id FROM strategy WHERE alias=\'' + alias + '\''
    cur.execute(query_string)
    data = cur.fetchall()

    if not data:
        strategy_id = None
    else:
        strategy_id = data[0][0]

    if 'con' not in kwargs.keys():
        con.close()

    return strategy_id


def get_strategy_info_from_alias(**kwargs):
    alias = kwargs['alias']
    con = msu.get_my_sql_connection(**kwargs)
    cur = con.cursor()

    query_string = 'SELECT * FROM strategy WHERE alias=\'' + alias + '\''
    cur.execute(query_string)
    data = cur.fetchall()

    if 'con' not in kwargs.keys():
        con.close()

    return {'id': data[0][0],
            'alias': data[0][1],
            'open_date': data[0][2],
            'close_date': data[0][3],
            'pnl': data[0][4],
            'created_date': data[0][5],
            'last_updated_date': data[0][6],
            'description_string': data[0][7]}


def load_trades_2strategy(**kwargs):

    trade_frame = kwargs['trade_frame']
    con = msu.get_my_sql_connection(**kwargs)

    trade_frame['strategy_id'] = [get_strategy_id_from_alias(alias=trade_frame['alias'][x],con=con) for x in range(len(trade_frame.index))]
    trade_frame['strike_price'] = trade_frame['strike_price'].astype('float64')

    now_time = dt.datetime.now()
    now_date = now_time.date()

    if 'trade_date' in kwargs.keys():
        trade_date = cu.convert_doubledate_2datetime(kwargs['trade_date'])
    else:
        trade_date = now_date

    column_str = "ticker, option_type, strike_price, strategy_id, trade_price, trade_quantity, trade_date, instrument, real_tradeQ, created_date, last_updated_date"
    insert_str = ("%s, " * 11)[:-2]

    final_str = "INSERT INTO trades (%s) VALUES (%s)" % (column_str, insert_str)

    column_names = trade_frame.columns.tolist()

    ticker_indx = column_names.index('ticker')
    option_type_indx = column_names.index('option_type')
    strike_price_indx = column_names.index('strike_price')
    trade_price_indx = column_names.index('trade_price')
    trade_quantity_indx = column_names.index('trade_quantity')
    instrument_indx = column_names.index('instrument')
    real_tradeQ_indx = column_names.index('real_tradeQ')
    strategy_id_indx = column_names.index('strategy_id')

    tuples = [tuple([x[ticker_indx],x[option_type_indx],
                     None if np.isnan(x[strike_price_indx]) else x[strike_price_indx],
                      x[strategy_id_indx],x[trade_price_indx], x[trade_quantity_indx],
              trade_date,x[instrument_indx], x[real_tradeQ_indx],now_time,now_time]) for x in trade_frame.values]

    msu.sql_execute_many_wrapper(final_str=final_str, tuples=tuples, con=con)

    if 'con' not in kwargs.keys():
        con.close()


def move_position_from_strategy_2_strategy(**kwargs):

    strategy_from = kwargs['strategy_from']
    strategy_to = kwargs['strategy_to']

    now_time = dt.datetime.now()

    con = msu.get_my_sql_connection(**kwargs)

    if 'con' not in kwargs.keys():
        kwargs['con'] = con

    if 'as_of_date' in kwargs.keys():
        as_of_date = kwargs['as_of_date']
    else:
        as_of_date = exp.doubledate_shift_bus_days()
        kwargs['as_of_date'] = as_of_date

    net_position_frame = get_net_position_4strategy_alias(alias=strategy_from, **kwargs)

    target_strategy_id = get_strategy_id_from_alias(alias=strategy_to, **kwargs)
    source_strategy_id = get_strategy_id_from_alias(alias=strategy_from, **kwargs)

    futures_position_frame = net_position_frame[net_position_frame['instrument'] == 'F']
    options_position_frame = net_position_frame[net_position_frame['instrument'] == 'O']

    futures_position_frame['trade_price'] = \
        [float(gfp.get_futures_price_4ticker(ticker=x,date_from=as_of_date,date_to=as_of_date,con=con)['close_price'][0]) for x in futures_position_frame['ticker']]

    if not options_position_frame.empty:
        options_position_frame['trade_price'] = options_position_frame.apply(lambda row: gop.get_options_price_from_db(ticker=row['ticker'],
                                                                    strike=row['strike_price'],
                                                                    option_type=row['option_type'],con=con,
                                                                    return_nan_if_emptyQ=True,
                                                                    settle_date=as_of_date)['close_price'][0], axis=1)

        net_position_frame = pd.concat([futures_position_frame,options_position_frame])
    else:
        net_position_frame = futures_position_frame

    column_names = net_position_frame.columns.tolist()

    ticker_indx = column_names.index('ticker')
    option_type_indx = column_names.index('option_type')
    strike_price_indx = column_names.index('strike_price')
    trade_price_indx = column_names.index('trade_price')
    trade_quantity_indx = column_names.index('qty')
    instrument_indx = column_names.index('instrument')

    column_str = "ticker, option_type, strike_price, strategy_id, trade_price, trade_quantity, trade_date, instrument, real_tradeQ, created_date, last_updated_date"
    insert_str = ("%s, " * 11)[:-2]

    final_str = "INSERT INTO trades (%s) VALUES (%s)" % (column_str, insert_str)

    tuples_target = [tuple([x[ticker_indx],x[option_type_indx],
                            None if np.isnan(x[strike_price_indx]) else x[strike_price_indx], target_strategy_id,
              x[trade_price_indx], x[trade_quantity_indx],
              as_of_date,x[instrument_indx], True,now_time,now_time]) for x in net_position_frame.values]

    msu.sql_execute_many_wrapper(final_str=final_str, tuples=tuples_target, con=con)

    tuples_source = [tuple([x[ticker_indx],x[option_type_indx],
                            None if np.isnan(x[strike_price_indx]) else x[strike_price_indx], source_strategy_id,
              x[trade_price_indx], -x[trade_quantity_indx],
              as_of_date,x[instrument_indx], True,now_time,now_time]) for x in net_position_frame.values]

    msu.sql_execute_many_wrapper(final_str=final_str, tuples=tuples_source, con=con)

    if 'con' not in kwargs.keys():
        con.close()


def get_open_strategies(**kwargs):

    con = msu.get_my_sql_connection(**kwargs)

    if 'as_of_date' in kwargs.keys():
        as_of_date = kwargs['as_of_date']
    else:
        as_of_date = int(tm.strftime('%Y%m%d'))

    cur = con.cursor()

    sql_query = 'SELECT * FROM futures_master.strategy WHERE close_date>=' + str(as_of_date) + ' and open_date<=' + str(as_of_date)

    cur.execute(sql_query)
    data = cur.fetchall()

    if 'con' not in kwargs.keys():
        con.close()

    return pd.DataFrame(data,columns=['id','alias','open_date','close_date','pnl','created_date','last_updated_date','description_string'])

def get_filtered_open_strategies(**kwargs):


    open_strategy_frame = get_open_strategies(**kwargs)
    open_strategy_frame['strategy_class'] = [conv.convert_from_string_to_dictionary(string_input=x)['strategy_class'] for
                                             x in open_strategy_frame['description_string']]

    return open_strategy_frame[open_strategy_frame['strategy_class'].isin(kwargs['strategy_class_list'])]


def select_strategies(**kwargs):

    con = msu.get_my_sql_connection(**kwargs)

    sql_query = 'SELECT * FROM futures_master.strategy WHERE '

    if 'open_date_from' in kwargs.keys():
        open_date_from = kwargs['open_date_from']
    else:
        open_date_from = 20170607

    sql_query = sql_query + ' open_date>=' + str(open_date_from)

    if 'open_date_to' in kwargs.keys():
        sql_query = sql_query + ' and open_date<=' + str(kwargs['open_date_to'])

    cur = con.cursor()

    cur.execute(sql_query)
    data = cur.fetchall()

    if 'con' not in kwargs.keys():
        con.close()

    return pd.DataFrame(data,columns=['id','alias','open_date','close_date','pnl','created_date','last_updated_date','description_string'])


def load_strategy_file(**kwargs):

    strategy_class = kwargs['strategy_class']
    report_date = kwargs['report_date']

    if strategy_class == 'futures_butterfly':
        output_dir = create_strategy_output_dir(strategy_class='futures_butterfly', report_date=report_date)
        strategy_output = pd.read_pickle(output_dir + '/summary.pkl')

    return strategy_output


def create_strategy_output_dir(**kwargs):

    strategy_class = kwargs['strategy_class']
    report_date = kwargs['report_date']

    strategy_output_folder = dn.get_directory_name(ext='strategy_output')

    if strategy_class == 'futures_butterfly':
        output_dir = strategy_output_folder + '/futures_butterfly/' + cu.get_directory_extension(report_date)
    elif strategy_class == 'curve_pca':
         output_dir = strategy_output_folder + '/curve_pca/' + cu.get_directory_extension(report_date)
    elif strategy_class == 'spread_carry':
         output_dir = strategy_output_folder + '/spread_carry/' + cu.get_directory_extension(report_date)
    elif strategy_class == 'vcs':
         output_dir = strategy_output_folder + '/vcs/' + cu.get_directory_extension(report_date)
    elif strategy_class == 'scv':
         output_dir = strategy_output_folder + '/scv/' + cu.get_directory_extension(report_date)
    elif strategy_class == 'ifs':
         output_dir = strategy_output_folder + '/ifs/' + cu.get_directory_extension(report_date)
    elif strategy_class == 'ics':
         output_dir = strategy_output_folder + '/ics/' + cu.get_directory_extension(report_date)
    elif strategy_class == 'os':
         output_dir = strategy_output_folder + '/os/' + cu.get_directory_extension(report_date)
    elif strategy_class == 'ts':
         output_dir = strategy_output_folder + '/ts/' + cu.get_directory_extension(report_date)
    elif strategy_class == 'itt':
         output_dir = strategy_output_folder + '/itt/' + cu.get_directory_extension(report_date)
    elif strategy_class == 'ocs':
         output_dir = strategy_output_folder + '/ocs/' + cu.get_directory_extension(report_date)
    elif strategy_class == 'ofs':
         output_dir = strategy_output_folder + '/ofs/' + cu.get_directory_extension(report_date)
    elif strategy_class == 'cot':
         output_dir = strategy_output_folder + '/cot/' + cu.get_directory_extension(report_date)
    elif strategy_class == 'arma':
         output_dir = strategy_output_folder + '/arma/' + cu.get_directory_extension(report_date)
    elif strategy_class == 'futures_directional':
         output_dir = strategy_output_folder + '/futures_directional/' + cu.get_directory_extension(report_date)

    if not os.path.exists(output_dir):
        os.makedirs(output_dir)

    return output_dir

File: c#/TA/TA/Strategy.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using MySql.Data;
using MySql.Data.MySqlClient;
using DatabaseConnection;
using System.Data;

namespace TA
{
    public class StrategyInfo
    {
        public int Id { set; get; }
        public DateTime OpenDate { set; get; }
        public DateTime CloseDate { set; get; }
        public double Pnl { set; get; }
        public DateTime CreatedDate { set; get; }
        public DateTime LastUpdatedDate { set; get; }
        public string DescriptionString { set; get; }
    }

    public static class Strategy
    {

        public static StrategyInfo GetStrategyInfoFromAlias(string alias, MySqlConnection conn)
        {
            StrategyInfo StrategyInfoOut = new StrategyInfo();

            string query = "SELECT * FROM strategy WHERE alias=\"" + alias + "\"";

            MySqlDataReader reader = null;
            MySqlCommand cmd = null;

            try
            {
                cmd = new MySqlCommand(query, conn);
                reader = cmd.ExecuteReader();

                while (reader.Read())
                {
                    StrategyInfoOut.Id = reader.GetInt32("id");
                    StrategyInfoOut.OpenDate = reader.GetDateTime("open_date");
                    StrategyInfoOut.CloseDate = reader.GetDateTime("close_date");

                    int PnlIndex = reader.GetOrdinal("pnl");

                    StrategyInfoOut.Pnl = reader.IsDBNull(PnlIndex) ? Double.NaN : reader.GetDouble("pnl");

                    StrategyInfoOut.CreatedDate = reader.GetDateTime("created_date");
                    StrategyInfoOut.LastUpdatedDate = reader.GetDateTime("last_updated_date");
                    StrategyInfoOut.DescriptionString = reader.GetString("description_string");
                }
            }

            finally
            {
                if (reader != null) reader.Close();
            }
            return StrategyInfoOut;
        }

        public static bool LoadTrades2Strategy(DataTable tradesTable, string alias, MySqlConnection conn)
        {
            String sqlStart = "insert into trades (ticker, option_type, strike_price, strategy_id, trade_price, trade_quantity, trade_date, instrument, real_tradeQ, created_date, last_updated_date) values ";
            List<string> tuples = new List<string>();

            foreach (DataRow Row in tradesTable.AsEnumerable())
            {
                tuples.Add(String.Format("({0},{1},{2},{3},{4},{5},{6},{7},{8},{9},{10})",
                                          "'" + Row.Field<string>("ticker") + "'",
                                         String.IsNullOrEmpty(Row.Field<string>("option_type")) ? "NULL" : "'" + Row.Field<string>("option_type") + "'",
                                         String.IsNullOrEmpty((Row.Field<decimal?>("strike_price").ToString())) ? "NULL" : Row.Field<decimal?>("strike_price").ToString(),
                                         TA.Strategy.GetStrategyInfoFromAlias(alias: Row.Field<string>("alias"), conn: conn).Id,
                                         Row.Field<decimal>("trade_price"),
                                         Row.Field<decimal>("trade_quantity"),
                                         "'" + DateTime.Now.ToString("yyyy-MM-dd") + "'",
                                         "'" + Row.Field<string>("instrument") + "'",
                                         Row.Field<bool>("real_tradeQ"),
                                          "'" + DateTime.Now.ToString("yyyy-MM-dd HH:mm:ss") + "'",
                                          "'" + DateTime.Now.ToString("yyyy-MM-dd HH:mm:ss") + "'"
                                         ));
            }

            return DatabaseConnection.Utils.MysqlExecuteManyWrapper(queryString: sqlStart, tuples: tuples, conn: conn);
        }

        public static bool LoadTrade2Strategy(string ticker,decimal trade_price,decimal trade_quantity,string instrument,string alias,MySqlConnection conn,bool real_tradeQ=true,string option_type=null,decimal?strike_price=null)
        {
            String sqlStart = "insert into trades (ticker, option_type, strike_price, strategy_id, trade_price, trade_quantity, trade_date, instrument, real_tradeQ, created_date, last_updated_date) values ";

            String ValuesString = String.Format("({0},{1},{2},{3},{4},{5},{6},{7},{8},{9},{10})",
                                          "'" + ticker + "'",
                                         String.IsNullOrEmpty(option_type) ? "NULL" : "'" + option_type + "'",
                                         String.IsNullOrEmpty((strike_price.ToString())) ? "NULL" : strike_price.ToString(),
                                         TA.Strategy.GetStrategyInfoFromAlias(alias: alias, conn: conn).Id,
                                         trade_price,
                                         trade_quantity,
                                         "'" + DateTime.Now.ToString("yyyy-MM-dd") + "'",
                                         "'" + instrument + "'",
                                         real_tradeQ,
                                          "'" + DateTime.Now.ToString("yyyy-MM-dd HH:mm:ss") + "'",
                                          "'" + DateTime.Now.ToString("yyyy-MM-dd HH:mm:ss") + "'");

            try
            {
                MySqlCommand cmd = new MySqlCommand(sqlStart + ValuesString, conn);
                cmd.ExecuteNonQuery();
            }
            catch
            {
                return false;
            }
            
            return true;
        }

        
        public static DataTable GetTrades(string alias, MySqlConnection conn)
        {
            string query = "SELECT tr.id, tr.ticker, tr.option_type, tr.strike_price, tr.trade_price, tr.trade_quantity, tr.trade_date, tr.instrument, tr.real_tradeQ " +
                 "FROM strategy as str INNER JOIN trades as tr ON tr.strategy_id=str.id " +
                 "WHERE str.alias=\"" + alias + "\"";

            MySqlCommand cmd = new MySqlCommand(query, conn);
            MySqlDataReader myReader = cmd.ExecuteReader();
            DataTable myDT = new DataTable();
            myDT.Load(myReader);
            return myDT;
        }

        public static DataTable GetNetPosition(string alias, MySqlConnection conn,DateTime? asOfDate = null)
        {
            DataTable Trades = GetTrades(alias,conn);

            DataTable netPosition = new DataTable();
            netPosition.Columns.Add("Ticker", typeof(string));
            netPosition.Columns.Add("OptionType", typeof(string));
            netPosition.Columns.Add("StrikePrice", typeof(double));
            netPosition.Columns.Add("Instrument", typeof(string));
            netPosition.Columns.Add("Qty", typeof(double));
            DataTable emptyPosition = netPosition.Copy();


            if (asOfDate.HasValue)
            {
                DataRow[] rows = Trades.Select("trade_date<= #" + ((DateTime)asOfDate).ToString("MM/dd/yyyy") + "#");
                if (rows.Count()==0)
                {
                    return emptyPosition;
                }
                Trades = rows.CopyToDataTable();
            }


            Trades.Columns.Add("fullTicker", typeof(string));

            foreach (System.Data.DataRow row in Trades.Rows)
            {
                if (row["instrument"].ToString() == "F")
                {
                    row["fullTicker"] = row["ticker"];
                }
                else if (row["instrument"].ToString() == "O")
                {
                    row["fullTicker"] = row["ticker"].ToString() + "_" +
                        row["option_type"].ToString() + "_" + row["strike_price"].ToString();
                }
            }

            var grouped = Trades.AsEnumerable().GroupBy(r => r["fullTicker"]).Select(w => new
            {
                fullTicker = w.Key.ToString(),
                ticker = w.First()["ticker"].ToString(),
                optionType = w.First()["option_type"].ToString(),
                instrument = w.First()["instrument"].ToString(),
                strikePrice = w.First()["strike_price"].ToString(),
                qty = w.Sum(r => decimal.Parse(r["trade_quantity"].ToString()))
            }).ToList();

            

            for (int i = 0; i < grouped.Count; i++)
            {
                netPosition.Rows.Add();
                netPosition.Rows[i]["Ticker"] = grouped[i].ticker;

                if (grouped[i].strikePrice.Length > 0)
                    netPosition.Rows[i]["StrikePrice"] = double.Parse(grouped[i].strikePrice);
                else
                    netPosition.Rows[i]["StrikePrice"] = double.NaN;


                netPosition.Rows[i]["OptionType"] = grouped[i].optionType;
                netPosition.Rows[i]["Instrument"] = grouped[i].instrument;
                netPosition.Rows[i]["Qty"] = grouped[i].qty;
            }

            DataRow[] NonEmptyPositions = netPosition.Select("qty<>0");

            if (NonEmptyPositions.Count()==0)
            {
                return emptyPosition;
            }

            return NonEmptyPositions.CopyToDataTable();
                
        }

        public static DataTable GetOpenStrategies(DateTime asOfDate,MySqlConnection conn)
        {
            string query = "SELECT * FROM futures_master.strategy WHERE close_date>=" + asOfDate.ToString("yyyyMMdd") + " and open_date<=" + asOfDate.ToString("yyyyMMdd");

            MySqlCommand cmd = null;
            MySqlDataReader reader = null;
            DataTable StrategyTable = new DataTable();

            try
            {
                cmd = new MySqlCommand(query, conn);
                reader = cmd.ExecuteReader();
                StrategyTable.Load(reader);
            }

            finally
            {
                if (reader != null) reader.Close();
            }
            return StrategyTable;
        }

        public static List<string> GetFilteredOpenStrategyList(DateTime asOfDate,MySqlConnection conn,string strategyClass=null, List<string> strategyClassList=null)
        {
            DataTable StrategyTable = GetOpenStrategies(asOfDate: asOfDate, conn: conn);
            StrategyTable.Columns.Add("StrategyClass", typeof(string));

            foreach (DataRow Row in StrategyTable.Rows)
            {
                Row["StrategyClass"] = Shared.Converters.ConvertFromStringToDictionary(stringInput: Row.Field<string>("description_string"))["strategy_class"];
            }

            if (!string.IsNullOrEmpty(strategyClass))
            {
                return (from x in StrategyTable.AsEnumerable()
                        where x.Field<string>("StrategyClass") == strategyClass
                        select x.Field<string>("alias")).ToList();
            }
            else
            {
                return null;
            }

            
        }

        public static string GenerateDbStrategyFromAlias(string alias,string descriptionString,MySqlConnection conn)
        {
            string aliasModified = alias;

            for (int i = 1; i < 11; i++)
			{
			 if (i>1)
             {
                 aliasModified = alias + "_" + i.ToString();
             }

             StrategyInfo Si = Strategy.GetStrategyInfoFromAlias(alias: aliasModified, conn: conn);

                if (Si.OpenDate==DateTime.MinValue)
                    break;
			}

            String sqlStart = "INSERT INTO strategy (alias, open_date, close_date, created_date, last_updated_date, description_string) values ";

            String ValuesString = String.Format("({0},{1},{2},{3},{4},{5})",
                "'" + aliasModified + "'",
                "'" + DateTime.Now.ToString("yyyy-MM-dd") + "'",
                "'" + (new DateTime(3000,1,1)).ToString("yyyy-MM-dd") + "'",
                 "'" + DateTime.Now.ToString("yyyy-MM-dd HH:mm:ss") + "'",
                  "'" + DateTime.Now.ToString("yyyy-MM-dd HH:mm:ss") + "'",
                   "'" + descriptionString + "'");

            MySqlCommand cmd = new MySqlCommand(sqlStart+ValuesString, conn);
            cmd.ExecuteNonQuery();
            return aliasModified;

        }

        public static bool CheckIfStrategyExist(string alias,MySqlConnection conn)
        {
            StrategyInfo Si = GetStrategyInfoFromAlias(alias: alias, conn: conn);
            return !(Si.OpenDate == DateTime.MinValue);

        }
    }
}
